waku/v2/node/peer_manager/peer_manager.nim

```diff
@@ -386,6 +394,41 @@ proc relayConnectivityLoop*(pm: PeerManager) {.async.} =
 
     await sleepAsync(ConnectivityLoopInterval)
 
+proc prunePeerStore(pm: PeerManager) {.async.}  =
+  while pm.started:
+    let numPeers = toSeq(pm.peerStore[AddressBook].book.keys).len
+    let capacity = pm.peerStore.capacity
+    if numPeers < capacity:
+      await sleepAsync(PrunePeerStoreInterval)
+      continue
+
+    debug "Peer store capacity exceeded", numPeers = numPeers, capacity = capacity
+    let peersToPrune = numPeers - capacity
+
+    # prune peers with too many failed attempts
+    var prunned = 0
+    for peerId in pm.peerStore[NumberFailedConnBook].book.keys:
+      if peersToPrune - prunned == 0:
+        break
+      if pm.peerStore[NumberFailedConnBook][peerId] >= pm.maxFailedAttempts:
+        pm.peerStore.del(peerId)
+        prunned += 1
+
+    # if we still need to prune, prune peers that are not connected
+    let notConnecteed = pm.peerStore.peers.filterIt(it.connectedness != Connected).mapIt(it.peerId)
```

Comment:
Checking my understanding: because maxConnections <= capacity we should always be able to reach our `peersToPrune` quota?

Reply:
exactly!

slightly related, since enforcing this maxConnections <= capacity is quite important added a unit test for this here a94325ab694bb38d34e655a7ba01a68dec76600b, otherwise, that wont be true.